File: src/csv_split.py
# convert csv into columnar format: output 1 csv per column
from misc import *

if len(args) < 2:
    err('usage:\n\t python3 csv_split.py [input csv filename]')

in_f, files = args[1], []  # input, outputs

with open(in_f, encoding="utf8", errors='ignore') as csvfile:
    csvreader = csv.reader(csvfile, delimiter=',')
    fields, n_fields, rng, ci, is_hdr = None, 0, 0, 0, True

    # for each row of csv
    for row in csvreader:
        row = [r.strip() for r in row]

        if is_hdr:
            # record field names
            fields, n_fields, rng = row, len(row), range(0, len(row))
            filenames, is_hdr = [in_f + '_' + f for f in fields], False

            # open a file per column
            for i in rng:
                print(row[i], filenames[i]) # print({row[i]: filenames[i] for i in rng})
            files = [open(filenames[i], 'wb') for i in rng]
            [files[i].write(fields[i].encode()) for i in rng]

        else:
            # write data row
            if len(row) != len(files):
                print("row", [row])
            print("ci", ci)
            [files[i].write(('\n' + row[i].rstrip()).encode()) for i in rng]
        ci += 1

    [files[i].close() for i in rng]  # close files
    print(ci, "rows written")

File: py/cpp/csv_split.cpp
#include"misc.h"

int main(int argc, char ** argv){
  long unsigned int n_error = 0;

  if(argc < 2){
    cout << "csv_split.cpp: split csv into columnar sets\n";
    err("usage:\n\tcsv_split.cpp [input file to split]");
  }
  string f_n(argv[1]);
  ifstream t(f_n);
  string s;

  int error;
  unsigned int i;
  FILE ** f = NULL;  // output files
  str newline("\n");
  vector<string> words;  // comma delimited chunks
  unsigned int n_f = 0;  // number of fields
  vector<string> field_names;  // names of the fields
  long unsigned int l_i = 0;  // line index

  while(getline(t,s)){
    error = false;
    words = split(s);

    if(l_i == 0){
      field_names = words;
      n_f = words.size();

      /* open a file for each field */
      f = (FILE **) alloc(sizeof(FILE *) * n_f);

      for0(i, n_f){
        str field_name(words[i]);
        std::replace(field_name.begin(), field_name.end(),'.', '_');
        str result("");
	std::remove_copy(field_name.begin(), field_name.end(), std::back_inserter(result), '*');
	field_name = result;
        string fn_i(string(f_n) + string("_") + field_name);
        cout << " +w " << fn_i << endl;
        f[i] = wopen(fn_i);
      }
      cout << "field_names: " << field_names << endl;
    }
    else{
      if(words.size() != n_f){
        cout << " l_i " << l_i << " " << words.size();
        cout <<	" n_f=" << n_f << " expected=" << words.size() << " " << words << endl;
        error = true;
        n_error ++;
        // exit(1);  (should this be commented out?)
      }
    }

    if(!error){
      for0(i, n_f){
        const char * word;

	if(l_i == 0){
          str field(words[i]);
          std::replace(field.begin(), field.end(), '.', '_');
          str result("");
          std::remove_copy(field.begin(), field.end(), std::back_inserter(result), '*');
          words[i] = result;
        }
        else{
          fprintf(f[i], "\n");
        }
        word = words[i].c_str();
        fwrite(word, strlen(word), 1, f[i]);
      }
    }

    if((++l_i) % 1000000 == 0){
      cout << words << endl;
    }
  }
  t.close();
  for0(i, n_f){
    fclose(f[i]);
  }
  cout << "n_error " << n_error << endl;
  cout << "done" << endl;
  return 0;
}
